contrib/shibuqam/examples/shibuqamoauth.nit

```diff
@@ -0,0 +1,157 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Server that implements an authentication bound to a `shibuqam` valid server.
+#
+# # Use as a web service
+#
+# ## Request
+#
+# Two GET fields
+#
+# * `redir`: A full URI that will be used to redirect the user once the authentication is done.
+# * `state`: A temporary string used to check that the callback is legitimate.
+#
+# On request, the service authenticate the user then redirect to the callback with the user informations.
+#
+# ## Response
+#
+# After the authentication, the response is a redirection to `redir` with the following GET arguments:
+#
+# * `id`: the user identifier. Its the code permanent for students. E.g. `"DOEJ00000000"`
+# * `name`: the full name. E.g. `"Doe, John"`
+# * `email`: the UQAM email. E.g. `"doe.john@courrier.uqam.ca"`
+# * `avatar`: the gravatar based on `email`
+# * `state`: the same state to check that the response is not forged.
+#
+# ## Errors
+#
+# If the request is badly formed or if the `redir` is not allowed, then there is no redirection and a text error message is send to the user.
+#
+# If there is a problem during the authentication, then there is a redirection but the GET fields are only `state` and `error`, where `error` is an error message.
+#
+# # To configure and execute
+#
+# ## Run the server
+#
+#    shibuqamoauth authorized_list [host] [port]
+#
+# `authorized_list` is a textfile that registers the list of accepted `redir`.
+#
+# Example:
+#
+# ~~~raw
+# https://example.com/foo
+# https://other.example.com/
+# ~~~
+#
+# To be authorized, a `redir` must have one of the line of `authorized_list` as a prefix.
+# For instance, the previous `authorized_list` accepts `https://example.com/foo` and `https://example.com/foo/bar`
+# but not `https://example.com/bar` nor `https://sub.example.com/foo`.
+#
+#
+# ## Install as a server
+#
+# A correct `shibuqam` reverse proxy must be configured (see `shibuqam` for details).
+# A full scenario is the following:
+#
+# * User: the human user and its browser that use the client website.
+# * Client: the web site that need authentication. It owns `https://client.example.com`
+# * Proxy: the public reverse proxy that does genuine Shibboleth authentication. It owns `https://server.example.com`
+# * Service: the private shibuqamoauth service. It owns `https://shibuqamoauth.example.com/`
+#
+# 1. User wants to login on client.
+# 2. Client chooses a random secret (state).
+#    Redirects the user to proxy as `https://server.example.com/login?redir=https://client.example.com/callback&state=secret`
```

Comment:
What is the actual purpose of `state`? Because, I can’t make a recommendation about it without knowing what security goal is intended.